### ISSUE TITLE
Investigate providing an API to combine coroutine builders and mutexes for a single line

### ISSUE THREAD

**Issue by ZakTaccardi** [Kotlin/kotlinx.coroutines#2449]
I might do something like the following:

```kotlin
val states: Flow<State> = scope.async {
    intentions.consumeAsFlow()
        .scan(initial = State()) { currentState, intention ->
            when (intention) {
                Intention.Unregister -> {
                    launch {
                        mutex.withLock {
                            send(Intention.Unregistering(isApiCallActive = true))
                            unregisterer.unregister()
                            send(Intention.Unregistering(isApiCallActive = false))
                        }
                    }
                    currentState
                }
                is Intention.Unregistering -> currentState.copy(
                    isUnregistering = intention.isApiCallActive
                )
            }
        }
        .stateIn(this)
}
    .flatten() // unwraps `Deferred<StateFlow<T>` into `Flow<T>` - see https://github.com/Kotlin/kotlinx.coroutines/issues/2446 
suspend fun send(intention: Intention) {
    intentions.send(intention)
}
data class State(
    val isUnregistering: Boolean = false
)
sealed class Intention {
    object Unregister : Intention()
    data class Unregistering(val isApiCallActive: Boolean) : Intention()
}
```

The mutex is needed because multiple API calls should not run in parallel, or the `State` could be incorrect. But, the nesting here is a little verbose:

```kotlin
launch {
    mutex.withLock {
        // ..
    }
}
```

Something more terse, like the following would be nice:

```kotlin
launch(mutex) {
  // ..  
}
```

**Comment by elizarov** [comment]
At the first sight, this pattern looks too domain-specific to be provided out-of-the-box in a general-purpose library. You can easily define this variant of `launch` yourself with just a few lines of code.
